[Lite Interpreter] Decouple JIT and autograd codes

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28900 [Lite Interpreter] Decouple JIT and autograd codes**

Decouple the JIT and autograd codes (and their dependencies). After this decoupling, the compressed torch mobile size is 548 KB total (comparing to 2.98 MB with full JIT).

Differential Revision: [D18226237](https://our.internmc.facebook.com/intern/diff/D18226237/)

Add "DECOUPLE_JIT" to make JIT related codes not compile when building mobile library.

iseeyuan committed:
[Lite Interpreter] Decouple JIT and autograd codes

Decouple the JIT and autograd codes (and their dependencies). After this decoupling, the compressed torch mobile size is 548 KB total (comparing to 2.98 MB with full JIT).

Differential Revision: [D18226237](https://our.internmc.facebook.com/intern/diff/D18226237/)

[ghstack-poisoned]

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1543,7 +1524,30 @@ struct CAFFE2_API ClassType : public NamedType {
+  void addMethod(Function* method);
+  Function* getMethod(const std::string& name) const;
+
+#ifndef DECOUPLE_JIT
```
Let's not change the API of this class based on a flag. My suggestion is only to change the implementation of these methods to throw exceptions. This prevents having to make modifications in as many places when adding a method that cannot be enabled for mobile.

iseeyuan replied:
Have removed the #ifndef in this header file and added implementations in the cpp file.

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1543,7 +1524,30 @@ struct CAFFE2_API ClassType : public NamedType {
+  void addMethod(Function* method);
+  Function* getMethod(const std::string& name) const;
+
+#ifndef DECOUPLE_JIT
```
The name DECOUPLE_JIT is not informative. I also don't see modifications to any build files, so I don't know if it is added correctly. Furthermore, I am surprised we don't already have a build flag indicating we are building for mobile. I am pretty sure C10 already has one.

iseeyuan replied:
The macro is currently for internal tests. Please refer to the internal diff D18226237 (which this PR is exported from) for testing cases. The reason for a different name is that the other name is for OSS mobile with full JIT. The two names may exist for a while till we fully move to lite interpreter.

iseeyuan replied:
The name has been changed to "USE_MOBILE_CLASSTYPE".

zdevito reviewed (commented):
I don't think the ifdef should appear in any header. Furthermore I don't see changes to build that would test that this works.

iseeyuan committed:
Update on "[Lite Interpreter] Decouple JIT and autograd codes"


Decouple the JIT and autograd codes (and their dependencies). After this decoupling, the compressed torch mobile size is 548 KB total (comparing to 2.98 MB with full JIT).

Differential Revision: [D18226237](https://our.internmc.facebook.com/intern/diff/D18226237/)

Add "DECOUPLE_JIT" to make JIT related codes not compile when building mobile library. 

[ghstack-poisoned]

iseeyuan committed:
Update on "[Lite Interpreter] Decouple JIT and autograd codes"


Decouple the JIT and autograd codes (and their dependencies). After this decoupling, the compressed torch mobile size is 548 KB total (comparing to 2.98 MB with full JIT).

Differential Revision: [D18226237](https://our.internmc.facebook.com/intern/diff/D18226237/)

Add "DECOUPLE_JIT" to make JIT related codes not compile when building mobile library. 

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -136,5 +139,15 @@ bool ClassType::isSubtypeOfExt(const TypePtr rhs, std::ostream* why_not) const {
 }
-
+#else
+bool ClassType::isSubtypeOfExt(const TypePtr rhs, std::ostream* why_not) const {
```
This should be implemented by just calling Type::isSubtypeOfExt. This reimplementation will bitrot.

iseeyuan replied:
If I remove this implementation but keep the declaration with override specifier in the header file, the compiler would complain. Any other idea?

zdevito replied:
Implement this by calling super:

`return Type::isSubtypeOfExt(rhs, why_not);`

iseeyuan replied:
Got it now. Thanks!

zdevito commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -92,19 +58,56 @@ size_t ClassType::addAttribute(
+  return cu;
+}
+
+#ifndef USE_MOBILE_CLASSTYPE
```
I don't see this set anywhere in this PR. How is this tested?

iseeyuan replied:
The macro is currently for internal tests. Please refer to the internal diff D18226237 (which this PR is exported from) for testing cases. It's more urgent for internal experiment at this time. OSS tests and cmake changes will be added after the internal experiments are done.

zdevito reviewed (approved):
The code looks good, but what tests that the mobile code is correct and disconnects the right parts of the JIT?

iseeyuan committed:
Update on "[Lite Interpreter] Decouple JIT and autograd codes"


Decouple the JIT and autograd codes (and their dependencies). After this decoupling, the compressed torch mobile size is 548 KB total (comparing to 2.98 MB with full JIT).

Differential Revision: [D18226237](https://our.internmc.facebook.com/intern/diff/D18226237/)

Add "DECOUPLE_JIT" to make JIT related codes not compile when building mobile library. 

[ghstack-poisoned]

iseeyuan committed:
Update on "[Lite Interpreter] Decouple JIT and autograd codes"


Decouple the JIT and autograd codes (and their dependencies). After this decoupling, the compressed torch mobile size is 548 KB total (comparing to 2.98 MB with full JIT).

Differential Revision: [D18226237](https://our.internmc.facebook.com/intern/diff/D18226237/)

Add "DECOUPLE_JIT" to make JIT related codes not compile when building mobile library. 

[ghstack-poisoned]

iseeyuan committed:
Update on "[Lite Interpreter] Decouple JIT and autograd codes"


Decouple the JIT and autograd codes (and their dependencies). After this decoupling, the compressed torch mobile size is 548 KB total (comparing to 2.98 MB with full JIT).

Differential Revision: [D18226237](https://our.internmc.facebook.com/intern/diff/D18226237/)

Add "DECOUPLE_JIT" to make JIT related codes not compile when building mobile library. 

[ghstack-poisoned]

iseeyuan committed:
Update on "[Lite Interpreter] Decouple JIT and autograd codes"


Decouple the JIT and autograd codes (and their dependencies). After this decoupling, the compressed torch mobile size is 548 KB total (comparing to 2.98 MB with full JIT).

Differential Revision: [D18226237](https://our.internmc.facebook.com/intern/diff/D18226237/)

Add "DECOUPLE_JIT" to make JIT related codes not compile when building mobile library. 

[ghstack-poisoned]
